nwb_conversion_tools/utils/specdatachunkiterator.py

```diff
@@ -0,0 +1,74 @@
+"""Authors: Cody Baker, Oliver Ruebel."""
+from typing import Iterable
+import numpy as np
+
+from hdmf.data_utils import AbstractDataChunkIterator, DataChunk
+
+
+class SpecDataChunkIterator(AbstractDataChunkIterator):
+    """DataChunkIterator that lets the user specify chunk shapes."""
+
+    def __init__(self, data: Iterable, full_shape: tuple, chunk_shape: tuple, dtype: str):
+        self.data = data
+        self.full_shape = full_shape
+        self.chunk_shape = chunk_shape
+        self.__curr_timestep = 0
+        self.__curr_channel_idx = 0
+        self.__dtype = DataChunk(data=np.empty(shape=(0, 0), dtype=dtype))
+
+    def __iter__(self):
+        """Return the iterator object"""
+        return self
+
+    def __read_block(self, start_time_idx, stop_time_idx, start_ch_idx, stop_ch_idx):
+        arr = self.data[start_time_idx:stop_time_idx, start_ch_idx:stop_ch_idx]
+        return arr
+
+    def __next__(self):
```

Comment:
I feel like this function would be using a `for` loop and `yield` instead of `if`s and `return`. I suppose it's just a stylistic difference, but I think the former is more common for these types of things

Reply:
The yield method is typically used when explicitly defining a data generator function, which would absolutely still be the case here if we extended coverage of this to data that is already in memory as an array (the current version assumes a well-functioning memmap).

Reply:
What do you mean?

Reply:
It wouldn't, that's just how we've used it for this purpose in the past. As I understand though we can't mess too much with the expected return structure of the `__next__` function, but we can probably control how it pulls the chunk from the original data source in read_blocks. I'll follow it up the chain of abstraction to confirm

Reply:
Can confirm, attempting to use a generator in the `__next__` function in place of the manual iterating with the `raise StopIteration` results in the usual long `hdmf` traceback terminating in

```
  File "D:\Users\Raven\anaconda3\envs\base2\lib\site-packages\hdmf\backends\hdf5\h5tools.py", line 839, in __exhaust_dcis
    if self.__write_chunk__(dset, data):

  File "D:\Users\Raven\anaconda3\envs\base2\lib\site-packages\hdmf\backends\hdf5\h5tools.py", line 1368, in __write_chunk__
    if isinstance(chunk_i.selection, tuple):

AttributeError: 'generator' object has no attribute 'selection'
```

Reply:
From the error it seems that your generator may not have returned a ``DataChunk``. In order to write the data HDMF needs to know both: 1) the data values to write and 2) the location to write to. That is what the``DataChunk``class does, it stores a chunk of data and a numpy selection to describe where the chunk belongs in the target array.

Reply:
This is the code for the attempt at using a data generator
```
    for curr_timestep in range(0, full_shape[0], chunk_shape[0]):
        for curr_channel_idx in range(0, full_shape[1], chunk_shape[1]):
            selection = np.s_[
                curr_timestep:(curr_timestep+chunk_shape[0]),
                curr_channel_idx:(curr_channel_idx+chunk_shape[1])
            ]
            data = data_in[selection]
            yield DataChunk(data=data, selection=selection)
```
The idea being that at each iteration it `yields` that particular `DataChunk`, but the bounds of the selector are much easier to specify within the `for` loops.

Reply:
I don't think you can have a yield here. Keep in mind, you are inside an iterator that itself iterates over the chunks, i.e., ``__next__`` is already being called inside a loop that stops iterating once you raise ``StopIteration``. I.e.,  in ``HDF5IO`` you have a loop of the form:

```
for chunk in mySpecDataChunkIterator:
    myHDF5Dataset[chunk.selection] = chunk.data
```

It looks a bit more complicated in ``HDF5IO`` because it has to manage a queue of possible multiple DataChunkIterators that need to be exhausted (see
https://github.com/hdmf-dev/hdmf/blob/c5f1bf7e476ce6538726db5fa84441c9ddde9957/src/hdmf/backends/hdf5/h5tools.py#L840-L848), but from the perspective of your ``SpecDataChunkIterator`` that is effectively what is happening. As such, the outer loop is outside of your direct control, since it is someone else (here ``HDF5IO``) that iterators over all elements (i.e., here ``DataChunk`` that you can produce). Placing a loop over all chunks inside ``__next__``, hence, means you now would have a nested loop iterating over all chunks inside a loop that tries to iterate over all chunks. Generally, using ``yield`` inside ``__next__`` does usually not make sense, because you now have an iterator that itself produces an iterator. This discussion seems related to this and may help: https://stackoverflow.com/questions/37929956/why-does-a-yield-from-inside-next-return-generator-object

Because of this, your loop counters (here ``curr_timestep`` and ``curr_channel``) need to be attributes on the  ``SpecDataChunkIterator`` class so you can keep track of the state of your loop, and then you need to update them inside your ``__next__``, just like you have it in the code right now 👍 

BTW, this also means if you don't reset ``self.__curr_timestep`` and ``self.__curr_channel_idx`` in ``__iter__`` or before you raise``StopIterator``, then the iterator is a on-time iterator only, i.e., if you want to reuse the same iterator again in another loop it will always raise StopIteration immediately. In this context here, I don't think this is an issue because the iterator is intended to be exhausted only once during write and not used again, I just figured I mention it just in case.

Reply:
The loop the way you have it in your example I think shouldwork if you placed it inside the ``__iter__`` function instead and leave ``__next__`` empty. Something like:

```
class SpecDataChunkIterator(AbstractDataChunkIterator):

   def __iter__(self):
        for curr_timestep in range(0, full_shape[0], chunk_shape[0]):
            for curr_channel_idx in range(0, full_shape[1], chunk_shape[1]):
                selection = np.s_[
                   curr_timestep:(curr_timestep+chunk_shape[0]),
                   curr_channel_idx:(curr_channel_idx+chunk_shape[1])
                ]
               data = data_in[selection]
               yield DataChunk(data=data, selection=selection)
```

Reply:
Great info, thanks for the details - I wasn't able to get it working by exclusively putting it in the `__iter__` (new issue was the call to next returned the `SpecDataChunkIterator` itself), but I was eventually able to get it to work by setting up the data generator as a function outside the class, making an internal attribute of the class equal to this, and having the `__next__` return the `next()` of that attribute.

Workflow can improve, I'm sure, but it is at least working. Makes the iteration bounds a lot easier to manage.

Reply:
Honestly, I think the original approach you used, where you updated of the loop variables inside ``__next__`` and keeping everything inside ``__next__`` seems like a cleaner design to me.

Reply:
Agreed - a lot changed in going to N-dimensions, though and the `__next__` is now back to doing what it more or less originally did but over generalized loops. The particular iteration of that loop still comes from an internal generator attribute (`self.chunk_idx_generator`), but it's also defined in a single line at the `__init__` stage so not that messy.

Reply:
A lot has changed from the early pushes of this; the internal structure of the GenericDataChunkIterator now more closely mimics the original Abstract design; chunks are extracted and set clearly in the `__next__`, shaping arguments set and pulled from respective private functions (including the ones used by the `hdmf`/`h5py` writing, like `recommended_chunk_shape`). It still uses an internal attribute that serves as the generating for the iterating permutations in N-dimensions, but I kinda like how that gets set at `__init__`.